_posts/new_post.markdown

```diff
@@ -0,0 +1,208 @@
+
+I don't have time or Devise knowledge required to dive deeper, but I [opened a GitHub issue](https://github.com/heartcombo/devise/issues/5716), so that anyone else with similar problems can find it, and that hopefully Devise fixes it in the next year or two.
+
+Now that could have been it. But you saw the title of the blog post already, so you know there's more. Of course there's more, look at the scrollbar position. 😂
+
+## Migrating away from Devise
+
+Now, Visualizer is not a huge app, but it's not a simple/tiny one either. I'm in no way pushing Devise to its limits, but I do use quite a lot of it: sign up flow, sign in flow, password reset flow, _omniauthable_ to Airtable, `authenticate` route constraint, and I also provide Doorkeeper OAuth flow.
+
+But, just for fun, how far can I push it with `rails generate authentication`? I [ran the generator](https://github.com/miharekar/visualizer/pull/112/commits/0b40abf3599e3a4b5c8a30655ca955e299bd582f) and found that it's pretty nice, yet some things are weirdly omitted: there's no _sign up flow_ and routes just use `resources` with no constraint which generates routes that controllers/views don't handle. Outside of that, it's pretty straight-forward: `Session` that belongs to `User`, with IP and User Agent persistence. It uses [`ActiveSupport::CurrentAttributes`](https://api.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html) which provides _thread-isolated per-request attributes_, which I was already familiar with, since _authentication-zero_ uses it as well.
+
+So after that, it was time for some [find & replace of `current_user` with `Current.user`](https://github.com/miharekar/visualizer/pull/112/commits/5ba5d5b35a96f8e3287a1423640173b6d1023672), `if current_user` with `if authenticated?`, `authenticate_user!` with `require_authentication`, changing to new sign in/up paths, and renaming `User`'s column from `encrypted_password` to `password_digest`. Yeah, _you're right_, it's not zero-downtime, but Postgres is pretty fast with these things, so it worked _just fine_ for my scale[^2]. Luckily both Devise and `has_secure_password` use `BCrypt::Password` under the hood, and since I haven't changed Devise's defaults, it _should just work_.
+
+And in testing it did mostly work, so then I was starting to wonder if there's a way to migrate users from Devise sessions/cookies seamlessly. After some googling I found that it stores `id` and first 30 characters of password salt in `session["warden.user.user.key"]` and `cookies.signed["remember_user_token"]`. The latter one stores one more thing, but it's irrelevant for our case. With that knowledge I was able to make this:
```

Comment:
Yes, it is correct. Devise, huh? 😄 

Honestly I didn't know what it was, since this value wasn't obvious: `1727516696.4554749`. After some deep dive into the code I figured it's `Time.now.utc.to_f.to_s`. So added to the text.